Fixes #8890 - Allow selection of plaintext "encryption" method for root ...

whefter opened the pull request:
...password

whefter committed:
Fixes #8890 - Allow selection of plaintext "encryption" method for root password

dLobatog commented:
@whefter Thanks for the contribution and sorry for the delay! Given the use case, why not just use base64 instead of plaintext? 

The `.po` and `.js` localization files are [auto generated](http://projects.theforeman.org/projects/foreman/wiki/Translating), you don't need to modify these, just the api docs.

Could you write a unit test for this? You can hunt us down in #theforeman-dev on IRC or the [mailing list](https://groups.google.com/forum/?fromgroups#!forum/foreman-dev) for help if you need. 

Again, sorry for the delay, other projects kept me a bit busy. [test] <- this is just a marker so that Jenkins picks up this PR and tests it.

stbenjam commented:
[test]

whefter commented:
My turn to apologize for the delay; had to find time to experiment with the test setup. Added unit test, fixed a bug, pushed the whole thing.

stbenjam commented:
[test]

stbenjam commented:
What takes base64 as a password option? This is exposed in the ENC verbatim, so seems like generally a bad idea.

stbenjam commented:
Ah, I see in the issue, Windows.  At the very least, there should probably be some code to _NOT_ expose the password in Puppet parameters when in an unhashed, plaintext format (https://github.com/theforeman/foreman/blob/develop/app/models/host/managed.rb#L345).

dLobatog commented:
[test] because of the latest enabled cops

whefter commented:
Fair point; I've added a check at the proposed location. Not sure if the whole Base64 thing warrants a quick re-write of the PasswordCrypt service since it's no longer, strictly speaking, only a crypt service.

dLobatog commented:
[test] , unfortunately this needs rebasing against the latest develop again, :bow: sorry it took us so long to get back to this, could you ping me here (@elobato) or on IRC #theforeman-dev whenever it's updated?

Thanks @whefter , really appreciate it

dLobatog commented on app/services/password_crypt.rb:
```diff
@@ -1,9 +1,11 @@
     raise Foreman::Exception.new(N_("Unsupported password hash function '%s'"), hash_alg) unless ALGORITHMS.has_key?(hash_alg)
-    passwd.crypt("#{ALGORITHMS[hash_alg]}#{SecureRandom.base64(6)}")
+    require 'base64'
+    # strict_base64 isn't available until Ruby 1.9.2
```
We don't support < 1.9.3 currently, so feel free to remove any alternative options and use strict if you want :smile: 

Also we usually put `require`s at the beginning of the file.

dLobatog commented:
@whefter I noticed you made some changes but you need to rebase against theforeman/foreman:develop. If you want to fix it, please add a remote 'upstream' to your repository pointing to the repo, then run `git pull --rebase upstream develop`, fix the conflicts, and then run `git rebase --continue` and resubmit.

Additionally the commits would have to be rebased into just one commit. Currently there are four of them so `git rebase -i HEAD~4` in your branch should rebase them.

Please also leave a comment pinging me '@eLobato' after you're done, otherwise Github doesn't notify me :person_frowning:  

Thanks again for your work, we really appreciate it!

dLobatog commented:
[test]

dLobatog commented:
@whefter It looks like [2 tests are failing](http://ci.theforeman.org/job/test_develop_pr_core/4418/) and [Rubocop](http://ci.theforeman.org/job/test_develop_pr_rubocop/2327/checkstyleResult/new/) too, could you take a look and ping me after?

Thanks, almost there!

whefter commented:
@eLobato Fixed the indentation issue (somehow, Kdiff thought that was a good idea), fixed the error and added a unit test to check that root_pw is indeed empty in the ENC when Base64 is selected (and present otherwise)

dLobatog commented:
[test]

dLobatog commented:
Merged as 331ff165939399787022e77ad17778e2ac39c148, thanks @whefter !

You'll notice it says Job result FAILURE, the tests passed and everything was mostly correct, but there were 3 trailing whitespaces in host_test.rb that I fixed on merge. Thanks again!
